libcaf_core/caf/detail/format.hpp

```diff
@@ -27,53 +27,40 @@
 namespace caf::detail {
 
 template <class T>
-inline constexpr bool is_formattable = is_complete<std::formatter<T, char>>;
-
-template <class T>
-inline constexpr bool fmt_needs_to_string
-  = !is_formattable<T>
-    && !std::is_same_v<
-      decltype(inspect_access_type<stringification_inspector, T>()),
-      inspector_access_type::none>;
-
-template <class T>
-constexpr T&&
-fmt_fwd(std::remove_reference_t<T>& arg,
-        std::enable_if_t<is_formattable<std::decay_t<T>>>* = nullptr) noexcept {
-  return static_cast<T&&>(arg);
-}
-
-template <class T>
-constexpr T&&
-fmt_fwd(std::remove_reference_t<T>&& arg,
-        std::enable_if_t<is_formattable<std::decay_t<T>>>* = nullptr) noexcept {
-  static_assert(!std::is_lvalue_reference_v<T>);
-  return static_cast<T&&>(arg);
-}
-
-template <class T>
-std::string
-fmt_fwd(std::remove_reference_t<T>& arg,
-        std::enable_if_t<fmt_needs_to_string<std::decay_t<T>>>* = nullptr) {
-  return deep_to_string(arg);
-}
-
-template <class T>
-std::string
-fmt_fwd(std::remove_reference_t<T>&& t,
-        std::enable_if_t<fmt_needs_to_string<std::decay_t<T>>>* = nullptr) {
-  static_assert(!std::is_lvalue_reference_v<T>);
-  return deep_to_string(arg);
+decltype(auto) fmt_fwd(T&& arg) {
+  if constexpr (std::is_default_constructible_v<
+                  std::formatter<std::decay_t<T>, char>>) {
```

Comment:
Why did you switch from `is_complete<std::formatter<T, char>>` to `is_default_constructible`?

Don't repeat yourself. We use `std::decay_t<T>` twice, give it an alias:

```cpp
  using arg_t = std::decay_t<T>;
```

Reply:
`is_complete` failed to compile for me. The class has a default template implementation, with an `static_assert(is_default_constructible<...>)` inside the class. This means the class is well defined, but because of the assert fails to compile. That is at least what I suspect, I can share the build log. Since the class uses `is_default_constructible` anyways, I'd leave it like this.